### PR TITLE
[Bugfix] Fix saving offloaded state dict

### PR DESCRIPTION
**Pull request by kylesayrs**
## Purpose ##
* Fix bug with saving the entire state dict of an offloaded model, even if it does not have a compressor
* Properly infer sparsity and quantization of offloaded models

## Changes ##
* Load the entire offloaded state dict before checking for a compressor and before inferring global sparsity
* Remove explicit override of `save_safetensors` kwarg, since the original `save_pretrained` function already defaults its value to `True`

## Testing ##
* Added test which throws error when trying to save an offloaded model without compression. This test fails on main but passes on this branch

Main
```
FAILED tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py::test_model_reload[True-torch_dtype3-False-cpu] - NotImplementedError: Cannot copy out of meta tensor; no data!
FAILED tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py::test_model_reload[True-torch_dtype5-False-cuda:0] - NotImplementedError: Cannot copy out of meta tensor; no data!
FAILED tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py::test_model_reload[True-torch_dtype6-True-cuda:0] - NotImplementedError: Cannot copy out of meta tensor; no data!
FAILED tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py::test_model_reload[True-torch_dtype7-True-cuda:0] - NotImplementedError: Cannot copy out of meta tensor; no data!
======================================= 4 failed, 4 passed, 12 deselected in 9.03s =======================================

```

This branch
```
============================================ 8 passed, 12 deselected in 8.89s ============================================

```

### CODE REVIEW

**Review by rahul-tuli — APPROVED** [review]

**Comment by kylesayrs** [comment]
I'm pretty sure this should be merged, but I want to fully test checkpoint loading before this is merged in

**Review by horheynm — DISMISSED** [review]

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py [hunk 1]
```diff
@@ -109,23 +113,16 @@ def save_pretrained_wrapper(
...
                 # model is not compressed or quantized, save as normal
                 original_save_pretrained.__get__(model, model_class)(
-                    save_directory, **kwargs
+                    save_directory, state_dict=state_dict, **kwargs
                 )
                 return
 
-            # if we've gotten to this point we have a config so we can run compression
-            # default safe serialization to True if not explicitly set
-            kwargs["safe_serialization"] = kwargs.get("safe_serialization", True)
```
Comment (dsikka) [inline src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:118]: This is already True by default?

**Review by dsikka — DISMISSED** [review]
LGTM. Just one confirmation question

File: src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py (hunk 1 above)
Comment (kylesayrs) [reply src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:118]: Yes, but let's make it explicit

Comment (kylesayrs) [reply src/llmcompressor/transformers/sparsification/compressed_tensors_utils.py:118]: [Yes](https://huggingface.co/docs/transformers/v4.44.2/en/main_classes/model#transformers.PreTrainedModel.save_pretrained), although there's no harm in keeping this line.

**Comment by dsikka** [comment]
@kylesayrs can you resolve the conflict

**Comment by kylesayrs** [comment]
I addressed the cuda errors by separating out GPU-dependent tests
There are other, new failures which are likely caused by upstream CT changes, but they are unrelated and outside of the scope of this PR, which only aims to address the newly added tests `test_model_reload`

File: tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py [hunk 2]
```diff
@@ -208,3 +210,58 @@ def test_quant_model_reload(format, dtype, tmp_path):
...
+    )
+
+    model_dict = get_state_dict_offloaded_model(model)
+    reloaded_dict = get_state_dict_offloaded_model(reloaded)
+    assert model_dict.keys() == reloaded_dict.keys()
+    for key in model_dict:
+        assert torch.equal(model_dict[key].cpu(), reloaded_dict[key].cpu())
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="requires gpu")
```
Comment (dsikka) [inline tests/llmcompressor/transformers/sparsification/test_compress_tensor_utils.py:254]: FYI - we do have `requires_gpu` decorators available but using unittest

**Review by dsikka — APPROVED** [review]

**Review by rahul-tuli — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Pull request merged**